src/ashell/comms-uart.c

```diff
@@ -322,31 +303,18 @@ void comms_write_buf(const char *buf, int len)
     if (len == 0)
         return;
 
-    struct device *dev = dev_upload;
-    uart_irq_tx_enable(dev);
-
+    uart_irq_tx_enable(dev_upload);
     data_transmitted = false;
-    uart_fifo_fill(dev, buf, len);
-
+    uart_fifo_fill(dev_upload, (const uint8_t *)buf, len);
```

Comment:
You're already busy waiting on data_transmitted so if that's anywhere close to okay, you should be fine to expand the loop to retry on the remaining bytes, until all your bytes are gone.

So basically:
```
while (len) {
    len -= uart_blah()
}

would maybe do it? Let's hear from our resident UART expert @jprestwo. :)